### PR TITLE
Ansible playbooks for hardware builds

### PR DESCRIPTION
**Pull request by erhudy**
- build bootstrap node
- PXE boot cluster nodes
- erase data disks
- set roles and converge cluster nodes
- other miscellaneous playbooks to come!

Also included:
- scripts to generate random cluster.txt files for testing, convert
  between cluster.txt and new cluster.yaml, and convert cluster.yaml to
  Ansible inventory files
- documentation on using the Ansible playbooks
- virtual hardware role to simulate hardware type (more to come here on
  using the Ansible playbooks on virtualized clusters)

This is based on work originally done by @caiush and @mihalis68.

### CODE REVIEW

**Comment by erhudy** [comment]
Filing the PR for tracking purposes, but don't merge it yet (there are some additional playbooks to add first and some additional work to be done to detect whether the playbook is dealing with hardware or virtualized nodes and use IPMI/VBoxManage calls accordingly).

**Comment by erhudy** [comment]
Removed label; additional playbooks are being worked on in feature/ansible-hardware-control. This is good to review as is.

**Comment by erhudy** [comment]
Since it was remarked that this pull request is really big, the only part of it that alters existing behavior is a change to the Vagrantfile that modifies how disks are attached to instances in order to more cleanly allow attaching 2 disks to a virtualized bootstrap node. The BCPC-Hardware-Virtual role is not referenced by anything but Ansible at the present time, and the scripts in `bootstrap/vbox_scripts` were never used by anything (I had put them there when I re-did the Vagrant bootstrap initially with the intent of fleshing them out as the hardware build path, but that never came to fruition).

**Comment by erhudy** [comment]
Just pushed some major changes to this stuff, including a few changes to the base cookbook. This needs to be tested carefully.

**Comment by erhudy** [comment]
Local Vagrant build works okay with the cookbook changes. Local Ansible build also works but exercises VirtualBox pretty hard, so I had a nice kernel panic that wrecked my cluster during a rechef. Need to go back and test again.

**Comment by erhudy** [comment]
Shoved up some other fixes and a whole lot of documentation changes. The hardware build guide is not done yet and needs to be stylistically adjusted, but you can probably bash your way through a hardware build by reading the local build guide.

**Comment by erhudy** [comment]
Crushed this down and I think it's ready to merge.

**Comment by erhudy** [comment]
Feedback for a follow-up PR to add some additional features:
* add a toggle for virtualized nodes to mount the apt mirror in via VirtualBox shared folder to save time and disk space (only available with **hardware_type=Virtual**)
* figure out a way to automatically configure the interfaces on the bootstrap node - preseeding would require a web server to be set up elsewhere to serve a preseed configuration file and PXE boot the bootstrap node, so Packer may be the best option here
* documentation probably needs some additional editing to be more bite-sized

**Comment by erhudy** [comment]
Additionally found a bug where if you attempt to converge the bootstrap without having synced the binary build files (e.g., `use_prebuilt_binaries=False`), Chef installation will fail because the packages aren't present. Will fix this in the follow-up PR.

File: docs/ansible_local_build_guide.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  * yaml
+  * Jinja2
+  * MarkupSafe
+* Download an Ubuntu server ISO from [http://www.ubuntu.com/download/server] to boot the bootstrap node from.
+
+Creating VMs
+---
+* Run `bootstrap/ansible_scripts/scripts/spawn_local_vms.sh` to build the VMs in VirtualBox (existing BCPC VMs will be deleted without mercy). This script will output the MAC addresses of the first network interface of each node.
+* Create `cluster.yml` in your staging directory, using `bootstrap/ansible_scripts/ansible-cluster.yml.example` as a model. For a local build, you should only need to replace the sample MAC addresses with the real ones.
+* Convert `cluster.yml` to `cluster.txt` with `bootstrap/ansible_scripts/scripts/cluster_yaml_to_inventory.py -t cluster.yml Test-Laptop-Ansible > cluster.txt`.
```
Comment (kelvk) [inline docs/ansible_local_build_guide.md]: The command should be `bootstrap/ansible_scripts/scripts/cluster_manifest_converter.py -t cluster.yaml Test-Laptop-Ansible > cluster.txt`

**Comment by erhudy** [comment]
Fixed error in documentation per @kelvk and force pushed.

**Comment by mihalis68** [comment]
I reviewed this multiple times and @kelvk tested it. Also provided feedback which is captured above which does not block this, merely adds some approachability considerations to be done later. On that basis I will now merge.

**Comment by erhudy** [comment]
Filing an issue to capture remaining concerns for the next PR.

**Pull request merged**